common/parse.go

```diff
@@ -114,66 +122,175 @@ func (jsonManager) ArrToMap(arr []*ajson.Node) ([]map[string]any, error) {
+	}
+
 	records, err := node.GetKey(key)
 	if err != nil {
 		return nil, err
 	}
 
 	arr, err := records.GetArray()
 	if err != nil {
-		return nil, ErrNotArray
+		return nil, formatProblematicKeyError(key, ErrNotArray)
 	}
 
 	return arr, nil
 }
 
-func (jsonManager) ArrSize(node *ajson.Node, key string) (int64, error) {
-	innerNode, err := node.GetKey(key)
+func (m jsonManager) ArrSize(node *ajson.Node, keys string) (int64, error) {
+	arr, err := m.GetArr(node, keys)
 	if err != nil {
 		return 0, err
 	}
 
-	if !innerNode.IsArray() {
-		return 0, ErrNotArray
+	return int64(len(arr)), nil
+}
+
+func (m jsonManager) GetStringWithDefault(node *ajson.Node, key string, defaultValue string) (string, error) {
+	result, err := m.GetString(node, key, true)
+	if err != nil {
+		return "", err
+	}
+
+	if result == nil {
+		return defaultValue, err
 	}
 
-	return int64(innerNode.Size()), nil
+	return *result, nil
 }
 
-func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (string, error) {
-	if optional && !node.HasKey(key) {
-		// null value in payload is allowed
-		return "", nil
+func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (*string, error) {
+	if !node.HasKey(key) {
+		if optional {
+			// null value in payload is allowed
+			return nil, nil // nolint:nilnil
+		} else {
+			return nil, createKeyNotFoundErr(key)
+		}
 	}
 
 	innerNode, err := node.GetKey(key)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+
+	if innerNode.IsNull() {
+		return nil, handleNullNode(key, optional)
 	}
 
 	txt, err := innerNode.GetString()
 	if err != nil {
-		return "", ErrNotString
+		return nil, ErrNotString
+	}
+
+	return &txt, nil
+}
+
+// GetNestedNode reaches into the JSON node by zooming via dot separated keys
+// Ex: keys = item.shipping.address => item has object with shipping key and so on.
+func (jsonManager) GetNestedNode(node *ajson.Node, keys []string) (*ajson.Node, error) {
+	var err error
+
+	// traverse nested JSON, use every key to zoom in
+	for _, key := range keys {
+		if !node.HasKey(key) {
+			message := fmt.Sprintf("%v; zoom=%v", key, strings.Join(keys, KeysZoomSeparator))
+
+			return nil, createKeyNotFoundErr(message)
+		}
+
+		node, err = node.GetKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !node.IsObject() {
+		return nil, ErrNotObject
+	}
+
+	return node, nil
+}
+
+func (jsonManager) ObjToMap(node *ajson.Node) (map[string]any, error) {
+	data, err := node.GetObject()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]any)
+	for k, v := range data {
+		result[k] = v
+	}
+
+	return result, nil
+}
+
+func handleNullNode(key string, optional bool) error {
+	if optional {
+		return nil
 	}
 
-	return txt, nil
+	return formatProblematicKeyError(key, ErrNullJSON)
+}
+
+func formatProblematicKeyError(key string, baseErr error) error {
+	return fmt.Errorf("problematic key: %v %w", key, baseErr)
+}
+
+func DotZoom(keys string) []string {
```

Comment:
This function appears to only be used in the test, and once you implement my suggestion of revising test input to be an array instead of a string, this function will no longer be necessary and can be removed.